# Tutorial Overview
1. __讀取檔案__.
2. __定義模型__.
3. __編譯模型__.
4. __擬合模型__.
5. __評估模型__.
6. __Summary__



## 數據集 [這裡下載](https://archive.ics.uci.edu/ml/datasets/pima+indians+diabetes):

- 資料屬性:


1. Number of times pregnant.  
2. Plasma glucose concentration a 2 hours in an oral glucose tolerance test 
3. Diastolic blood pressure (mm Hg) 
4. Triceps skin fold thickness (mm) 
5. 2-Hour serum insulin (mu U/ml) 
6. Body mass index (weight in kg/(height in m)^2) 
7. Diabetes pedigree function 
8. Age (years) 
9. Class variable (0 or 1)


- 數據格式

6, 148, 72, 35, 0, 33.6, 0.627, 50, 1

1, 85, 66, 29, 0, 26.6, 0.351, 31, 0

8, 183, 64, 0, 0, 23.3, 0.672, 32, 1

1, 89, 66, 23, 94, 28.1, 0.167, 21, 0

### 1.讀取檔案.

In [2]:
from keras.models import Sequential
from keras.layers import Dense
import numpy as np

# fix random seed for reproducibility
np.random.seed(87)

# numpy讀檔函式 laodtxt
dataset = np.loadtxt("pima-indians-diabetes.data.txt", delimiter=",")

# 分割檔案為，輸入/標籤
X = dataset[:,0:8]
Y = dataset[:,8]

Using TensorFlow backend.


In [9]:
#此為測試區！自行測試數據
X.shape

(768, 8)

### 2.定義模型.

- Keras疊積木時間。


1. 宣告序列輸入
2. 輸入8個，進入10個神經元隱藏層
3. 10個再進入16個隱藏層
4. 輸出1個位元

    - keras神經網路權重weights初始化預設是，均勻分布0 ~ 0.05。
    

- Q : input_dim = 8 為什麼輸入神經元要設定成為8 ?
- Q : 為什麼Activation 要用relu，但是最後卻是sigmoid？

- A : 因為每次輸入數據有八個不同數值，sigmoid微分爆炸/消失有關， google **gradient vanish**

In [6]:
model = Sequential()
model.add(Dense(10, input_dim=8, activation='relu')) 
model.add(Dense(16, activation='relu')) 
model.add(Dense(1, activation='sigmoid'))

### 3.編譯模型.

1. 我們使用的損失函數是：二位元交互訊息。盡可能最小損失。
2. 優化方式： Adam
3. 度量子：正確性。想辦法提高。

Adam: A Method for Stochastic [Optimization]( http://arxiv.org/abs/1412.6980)

In [7]:
model.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])

### 4.擬合模型.

- Hint :從768筆資料抽出幾筆(batch size)去做訓練，重複一次叫做一個epoch。

- 參數可自行調整

In [8]:
model.fit(X, Y, epochs=150, batch_size=10)

Epoch 1/150
768/768 [==============================] - 0s - loss: 2.7916 - acc: 0.6693     
Epoch 2/150
768/768 [==============================] - 0s - loss: 0.8642 - acc: 0.6263     
Epoch 3/150
768/768 [==============================] - 0s - loss: 0.6582 - acc: 0.6549     
Epoch 4/150
768/768 [==============================] - 0s - loss: 0.6362 - acc: 0.6719     
Epoch 5/150
768/768 [==============================] - 0s - loss: 0.6332 - acc: 0.6471     
Epoch 6/150
768/768 [==============================] - 0s - loss: 0.6213 - acc: 0.6862     
Epoch 7/150
768/768 [==============================] - 0s - loss: 0.6170 - acc: 0.6810     
Epoch 8/150
768/768 [==============================] - 0s - loss: 0.6111 - acc: 0.6810     
Epoch 9/150
768/768 [==============================] - 0s - loss: 0.6086 - acc: 0.6888     
Epoch 10/150
768/768 [==============================] - 0s - loss: 0.6045 - acc: 0.6823     
Epoch 11/150
768/768 [==============================] - 0s - loss: 0.6027 - acc

768/768 [==============================] - 0s - loss: 0.5094 - acc: 0.7487     
Epoch 90/150
768/768 [==============================] - 0s - loss: 0.5064 - acc: 0.7617     
Epoch 91/150
768/768 [==============================] - 0s - loss: 0.5079 - acc: 0.7487     
Epoch 92/150
768/768 [==============================] - 0s - loss: 0.5108 - acc: 0.7500     
Epoch 93/150
768/768 [==============================] - 0s - loss: 0.5135 - acc: 0.7396     
Epoch 94/150
768/768 [==============================] - 0s - loss: 0.5065 - acc: 0.7383     
Epoch 95/150
768/768 [==============================] - 0s - loss: 0.5043 - acc: 0.7500     
Epoch 96/150
768/768 [==============================] - 0s - loss: 0.5094 - acc: 0.7396     
Epoch 97/150
768/768 [==============================] - 0s - loss: 0.5072 - acc: 0.7409     
Epoch 98/150
768/768 [==============================] - 0s - loss: 0.4976 - acc: 0.7578     
Epoch 99/150
768/768 [==============================] - 0s - loss: 0.5040 - acc: 0.

### 5.評估模型

我們訓練神經網路，在整個數據裡面，只會顯示出我們訓練的好壞，透過正確性來看，但是卻無法得知，訓練好的模型透用在新的數據上表現如何。
簡單的做法就是，我們把數據拆解成，訓練/測試兩個數據，我們在這先把訓練好的模型，套用原先的數據，看看表現如何。

- Q: 可以把原先數據再套用上去嗎？

In [12]:
scores = model.evaluate(X, Y)
print("\n%s: %.2f%%" % (model.metrics_names[1], scores[1]*100))

 32/768 [>.............................] - ETA: 0s
acc: 77.34%


### 6. Summary

1. 讀取檔案 numpy.loadtxt
2. 定義模型 model = Sequential(), model.add()...
    
3. 編譯模型，損失函數，優化方式，度量方式
4. 擬合 fit(X, Y, epochs=150, batch_size=10)
5. 評估模型evaluate(X, Y)